[iOS] 성공 모크 스텁 데이터로 메인화면에 데이터를 입혔습니다.

ehgud0670 opened the pull request:
이슈 #7 을 해결하였습니다.

* 성공 모크 스텁 객체를 이용해 메인 화면에 데이터를 입혔습니다.
  * `CategoryViewModel`이 `CategoryHeader` 모델 객체를 프로퍼티로 갖고, `CategoryHeaderView`에 데이터를 바인딩하도록 하였습니다. 
  * `CategoryViewModel`이 `ProductViewModel`을 배열로 갖고, `ProductViewModel` 각각이 `ProductCell`에 데이터를 바인딩하도록 하였습니다. 
  * `CategoryViewModels`가  프로퍼티로 `CategoryViewModel` 배열을 가지고 있고, 관련 동작을 하도록 하였습니다.

* json데이터를 파일로 저장하고 Bundle.main.url() 을 사용하여 데이터를 불러오도록 하였습니다.
* reloadSections를 사용하여 해당 index의 섹션만 업데이트하도록 기능 구현하였습니다.
> 고민인 점

EventBadgeStackView 객체에 어떤 뷰도 추가하지 않은 채로 ProductCell 객체에 오토레이아웃을 적용하면, 이후에 벳지 레이블 객체를 추가할때 레이블이 보이지 않는 이슈가 있었습니다. 
따라서 저는 EventBadgeStackView 클래스에 intrinsicContentSize를 오버라이딩하여 벳지 레이블의 사이즈만큼 적용하였고, 이후에 뱃지 레이블 객체를 추가해도 레이블이 보여질 수 있게끔 하였는데 이게 맞는 방법인지 모르겠습니다...

EventBadgeStackView.swift

```swift
    override var intrinsicContentSize: CGSize {
        let label = EventBadgeLabel()
        return label.intrinsicContentSize
    }
```

ehgud0670 committed:
[iOS] Refactor: intrinsicContentSize 를 오버라이딩하여 재적용함
		이전의 widthAnchor는 필요없어져서 삭제함

ehgud0670 committed:
[iOS] Refactor: 상품 이미지가 왼쪽 테두리에서 더 나오도록 constant를 줌

associate with #7

ehgud0670 committed:
[iOS] Chore: 성공 스텁 데이터로 사용할 json 파일들 추가함

ehgud0670 committed:
[iOS] Feat: 1. stub json 파일 추가하고 json 데이터를 반환하는 정적 메소드 구현함
	    2. 데이터가 잘 리턴되는지 테스트 코드 작성함

ehgud0670 committed:
[iOS] Feat: protocol NetworkManagable을 만들고 이것을 채택한 NetworkManager를 구현함

associate with #7

ehgud0670 committed:
[iOS] Refactor: request에 httpMethod와 httpBody를 할당함
	이유 : 원래 해야 하는 건데 하지 않았다;;
associate with #7

ehgud0670 committed:
[iOS] Feat: 모크 성공 객체들을 생성하고 핸들러로 넘어온 데이터가 널이아닌지 테스트하는
	   테스트 코드 작성함. 추후에 모델 객체가 잘 처리되는 테스트할 것임
associate with #7

ehgud0670 committed:
[iOS] Feat: categoryURLs를 request하여 받은 응답을 처리하는 모델 객체 생성 및
	    잘 처리되었는지 테스트 코드 작성함
	    => 잘 처리됨

associate with #7

ehgud0670 committed:
[iOS] Feat: 1. Category를 request할때 응답받은 데이터를 처리할 Codable 객체 생성함
	    2. 테스트 코드를 작성하여 잘 처리되었는지 확인함

ehgud0670 committed:
[iOS] Refactor: invalidURL인 경우를 처리하는 코드 삭제함
	이유: 모두 Mock, Stub를 이용하여 테스트하므로 invalidURL인지 검사하는 것은 의미가 없다고 생각하여 삭제함
associate with #7

ehgud0670 committed:
[iOS] Feat: 상품 상세 데이터 요청에  대한 응답을 처리할 Codable 객체 생성 및
	    테스트 코드로 Codable 객체가 잘 생성됨을 확인함
associate with: #7

ehgud0670 committed:
[iOS] Feat: viewDidLoad에 categoryurls를 받는 메소드 구현함
	    이후에 이 url로 category 데이터를 요청하고 응답받는 메소드를 만들 예정이다.
associate with #7

ehgud0670 committed:
[iOS] Feat: 카테고리 데이터들을 요청하고 받는 메소드 기능 구현함
associate with #7

ehgud0670 committed:
[iOS] Refactor: category 모델을 만들어 메소드가 응답 객체를 category 모델화 하도록 함
		또 다른 곳도 알맞게 메소드명을 바꾸거나 재적용함

ehgud0670 committed:
[iOS] Feat: 데이터를 로드해서 모델 객체들을 만들면  모델 배열에 순서에 맞게 넣어지도록 기능 구현함

ehgud0670 committed:
[iOS] Chore: 호출되는 순서에 맞게 메소드 재배치함
associate with: #7

ehgud0670 committed:
[iOS] Refactor: 기존 프로퍼티인 모델 객체 배열을 지우고, 헤더뷰모델, 상품들 뷰모델을 각각 만들어서
	프로퍼티로 각각 배열을 만든후 데이터를 넣도록 함

ehgud0670 committed:
[iOS] Refactor: 애초에 Category 모델로 만드는게 아니라 각각의 뷰 모델을 만들도록 수정함
		필요없는 Category 모델 삭제함
associate with #7

ehgud0670 committed:
[iOS] Refactor: 기존의 데이터소스 객체를 없애고, 뷰컨이 데이터소스 역할을 하도록 기능 구현함
	이유: 일단 역할을 뷰컨이 모두 하게 하고 이후에 따로 데이터소스, 델리게이트 객체를 만들어서
	      역할을 분리하도록 할 것이다.

ehgud0670 committed:
[iOS] Fix: 헤더의 수와 상품의 수가 잘 나오도록 코드 수정함

ehgud0670 committed:
[iOS] Refactor: 처음 로드 받기 전인 경우 개수를 모두 0을 리턴하도록 기능 수정함

ehgud0670 committed:
[iOS] Feat: 각각의 뷰모델이 모델을 뷰에 바인딩하도록 코드를 작성함

associate with #7

ehgud0670 committed:
[iOS] Refactor: 불필요하다고 판단한 스택뷰들 삭제하고 레이블들 오토레이아웃 직접 지정함

ehgud0670 committed:
[iOS] Fix: 이전 커밋에선 이벤트 벳지가 보이지 않았지만 temp를 추가하고
	  안보이게 함으로써 임시방편으로 보이게함
	마음에 안들어서 추후 더 나은 방법을 모색할 것임

ehgud0670 committed:
[iOS] Refactor: constraint constant 재조정함

ehgud0670 committed:
[iOS] Feat: 일단은 뷰모델이 변화되면 모두 다시 업데이트하도록 기능 구현함

ehgud0670 committed:
[iOS] Refactor: 필요없어진 스택뷰들 삭제함

ehgud0670 committed:
[iOS] Feat: prepareForReuse를 사용해서 cell 초기화함
      Refactor: stackView intrinsicContentSize를 오버라이딩하여 처음부터 height를 event라벨만큼 설정해줌

ehgud0670 committed:
[iOS] Style: 높이를 주기 위한 텍스트이므로 문자열 내용을 "높이"로 줌

ehgud0670 committed:
[iOS] Refactor: 뷰 관련 데이터 가공을 뷰모델이 하게 하고,
                    기존 분기문을 뷰의 메소드에서 처리되도록 리팩토링함
associate with #7

ehgud0670 committed:
[iOS] Chore: 개행 정리함
associate with #7

ehgud0670 committed:
[iOS] Refactor: CategoryHeaderViewModel은 지우고 CategoryViewModel을 만들어서 재적용함
	이유: 이전의 코드는 HeaderViewModel과 ProductViewModel이 따로 분리되어 있어서
		자료구조가 복잡했다.
	      하지만 이제 CategoryViewModel이란 것이 ProductViewModel 배열을 갖게하여
	     좀 더 자료구조가 유연해지고 응집성이 강하도록 하였다.

ehgud0670 committed:
[iOS] Refactor: viewModel 배열을 관리하는 ViewModels 객체를 만들고 배열에 변화가 있으면
		노티를 보내 reloadData를 하도록 기능 구현함

	이유: 상위 모듈은 단순하게, 하위모듈은 복잡하게. 크,,,취한다

associate with #7

ehgud0670 committed:
[iOS] Refactor: DataSource 객체를 따로 만들어서 책임을 분리함,
		클로저를 이용한 바인딩으로 모델을 주입함

	이유: 뷰컨트롤러가 massive 해지는 걸 방지하기 위해 역할을 나누었다.
		클로저를 사용해 모델 값을 넣는것도 하고 싶었다.

associate with #7

ehgud0670 committed:
[iOS] Refactor: asyncAfter 문 삭제함
	이유: 비동기적으로 데이터가 잘 바인딩되는지 테스트하기 위한 임시 조치이므로
		이제 실제 서버로부터 데이터를 로드할 것이기 때문에 삭제한다.

ehgud0670 committed:
[iOS] Refactor: 1. UseCase객체를 만들어서 CategoryViewModel이라는
	유의미한 모델 객체를 핸들러에 던지도록 기능 수정함
	2. 함수가 하나의 책임만을 하도록 함수를 분리함
	   => url배열을 일단 request 해서 url 배열을 받으면
	      url배열을 가지고 usecase를 설정하는 메소드를 호출하도록 함

ehgud0670 committed:
[iOS] Chore: 함수 호출 순서에 맞게 함수 위치 순서를 정리함
associate with #7

ehgud0670 committed:
[iOS] Refactor: configureUsecase 메소드에 다시 urls를 요청하는 함수를 합침
	이유: 분리하다보니 viewDidLoad에서 함수 구현이 가독성이 떨어져보였다.
		그리고 configureUseCase에서 urls를 요청하는 것도 하나의 책임으로 할 수 있다고 생각하기에
		다시 합쳤다.
associate with #7

ehgud0670 committed:
[iOS] Style: 통일성을 주기 위해 이름에서 Food와 Menu를 지움
associate with #7

ehgud0670 committed:
[iOS] Refactor: reloadSections 을 사용하여 해당 index의 section만 업데이트하도록 기능 수정함

godrm commented on iOS/Sidedish/Sidedish/DataSources/CategoryTableViewDataSource.swift:
```diff
@@ -0,0 +1,53 @@
+            let salePriceText = productViewModel.text(price: product.sale_price)
+            productCell.configure(normalPriceText: normalPriceText,
+                                  salePriceText: salePriceText,
+                                  unitText: productViewModel.unitText)
```
normalPriceText 가 없으면 아예 가격정보를 표시 안하는건가요? -ㅁ- 
여기 productCell.xxx() 형태가 많은데 product를 어차피 값만 있는거라면 통째로 넘기는건 어때요?

ehgud0670 replied:
> normalPriceText

넵, 경우의 수가 
     1) 정상가격과 할인가격이 모두 있는 경우 
     2) 정상가격만 있는 경우 
라서 normal_price 는 반드시 있는 데이터이고, normal_price(Int)에서 normalPriceText(String)을 성공적으로 가져오는 경우만 
가격정보를 표시하고,  normalPriceText가 nil인 경우는 return 하도록 하였습니다. 
반드시 있는 데이터여서 guard를 썼는데 안괜찮은 건가요 ㅜㅜ

> product

아 근데, 가격의 경우에는 Int 타입을 뷰모델 객체가 String 타입으로 바꾼 뒤에 cell에 데이터를 입히는 것인데 product 모델을 통째로 넘기는게 될지 모르겠습니다 ㅜㅜ

godrm replied:
이상하네요. 이 글은 댓글이 안 달렸네요. 
꼭 서버에서 받은 Product를 뷰에 전달해야 하는 건 아니죠. 
가격정보가 없으면 셀 전체를 안보여주는 게 맞는걸까요?

godrm commented on iOS/Sidedish/Sidedish/Network/Mocks/MockCategorySuccessStub.swift:
```diff
@@ -0,0 +1,18 @@
+                         httpMethod: HTTPMethod, httpBody: Data?,
+                         completionHandler: @escaping (Data?, URLResponse?, Error?) -> ()) throws {
+        guard let data = Data.jsonData(forResource: "SuccessCategoryResponseStub") else { return }
+        completionHandler(data, nil, nil)
```
실패할 때 핸들러가 따로 없는게 아쉽네요.

ehgud0670 replied:
넵, 다음 이슈에서 실패하는 경우의 핸들러를 만들어보도록 하겠습니다!

ehgud0670 replied:
넵. 다음 이슈에서 만들어보도록 하겠습니다! ㅎㅎ

godrm commented on iOS/Sidedish/Sidedish/Network/Mocks/MockProductDetailSuccessStub.swift:
```diff
@@ -0,0 +1,18 @@
+
+import Foundation
+
+struct MockProductDetailSuccessStub: NetworkManagable {
```
Mock과 Stub 을 다 붙이고 있는데 어떤 의도가 있는걸까요?
https://blog.pragmatists.com/test-doubles-fakes-mocks-and-stubs-1a7491dfa3da 자료를 참고하세요. 
![mock](https://user-images.githubusercontent.com/278988/80354659-00f78880-88b2-11ea-969f-ce22a87b1cd6.png)
![stub](https://user-images.githubusercontent.com/278988/80354673-048b0f80-88b2-11ea-9ab4-c41e6574bd90.png)

ehgud0670 replied:
Mock객체는 어떠한 객체를 따라하는 가짜 객체이고, Stub 데이터는 가짜 데이터라고 알고 있습니다. 
저는 한 Mock 객체가 네트워크를 성공적으로 연동한 경우에 Stub 데이터를 넘기는 역할인걸 표현하고 싶어서  모크 뒤에 스텁이라고 이름을 지었습니다. 
보내주신 자료 공부하도록 하겠습니다. 감사합니다 ㅎㅎ

ehgud0670 replied:
넵 혼용하지 않고 사용하지 않도록 하겠습니다. 적용하였습니다 ㅎㅎ

godrm commented on iOS/Sidedish/Sidedish/Network/Stubs/DataExtensions.swift:
```diff
@@ -0,0 +1,17 @@
+import Foundation
+
+extension Data {
+    static func jsonData(forResource name: String?) -> Data? {
```
매개변수까지 String? 옵셔널일 필요가 있나요? 그럼 호출하지 않아도 되는거 아닐까요?

ehgud0670 replied:
아 넵 맞습니다 ㅎㅎ
반영하였습니다!

godrm commented on iOS/Sidedish/Sidedish/UseCases/CategoryViewModelUseCase.swift:
```diff
@@ -0,0 +1,30 @@
+
+struct CategoryViewModelUseCase {
+    static func makeCategoryViewModel(from urlString: String,
+                                   with manager: NetworkManagable,
```
makeCategoryViewModel() 유즈케이스도 CategoryViewModel을 만드는 게 목적이 아니지 않나요?
서버에 요청하면 결과로 모델이 만들어지는 게 아닐까요?
유즈케이스 수준에서는 모델만 만들고 뷰 모델을 만드는 건 그 다음 단계가 어떨까요
뷰 모델은 UIKit에 의존하거나 뷰 정보를 알고 있는 상태라서 Presenter나 데이터소스, 뷰컨트롤러 수준에서만 알 수 있는 데이터입니다.

ehgud0670 replied:
아 넵 감사합니다! 모델만 유스케이스가 핸들러로 넘기고 그 다음 단계에서 뷰모델을 만들도록 하겠습니다!

ehgud0670 replied:
반영하였습니다!

godrm commented on iOS/Sidedish/Sidedish/ViewControllers/CategoryViewController.swift:
```diff
@@ -0,0 +1,141 @@
+                guard let categoryViewModels = self.categoryViewModels,
+                    let categoryViewModel = categoryViewModels.categoryViewModel(at: section),
+                    let productViewModel = categoryViewModel.productViewModel(at: row) else { return nil }
+                return productViewModel
```
여기 핸들러들은 왜 필요한지 모르겠어요. 그냥 데이터 소스 내부에 있으면 되는 코드 같은데요.

ehgud0670 replied:
아 저는 `CategoryViewModels` 객체가 뷰모델 데이터들을 가지고 있는 경우라, 이 데이터들을 데이터소스가 사용하기 위해서 핸들러들을 사용하였습니다. 
말씀해주신게 아예 데이터소스가 뷰모델을 갖는게 어떻냐는 말씀이신가요?

godrm replied:
DataSource와 categoryViewModels 가 나눠져 있는 것도 이상하다고 볼 수 있죠. categoryViewModels 객체가 UITableViewDataSource를 채택하고 있어도 되는거 아닐까요?
뷰 모델이 뷰를 그리기 위한 데이터를 갖고 있어야 하는데 테이블뷰 뷰모델은 데이터소스가 갖고 있어야 하니까요. 저번에도 그랬던 것 같은데 둘이 같은 일을 하고 있는데 이름만 나눠놓고 따로 만들고 있는 것 같아서요.

ehgud0670 replied:
아 테이블뷰 뷰모델이 데이터소스를 갖고 있어야 하고, 뷰 모델이 뷰를 그리기 위한 데이터를 갖고 있어야 하므로 둘이 같은 일을 하는 것이군요. 넵 그럼 categoryViewModels객체가 UITableViewDatasource를 채택하도록 하겠습니다. 지난번에는 데이터소스가 뷰 모델 로서 역할도 했는데 의미를 명확히 몰라서 이번에 나눠보았었는데 이제야 의미를 알 것 같아요.
감사합니다! ㅎㅎ

ehgud0670 replied:
반영하였습니다!

godrm commented on iOS/Sidedish/Sidedish/ViewControllers/CategoryViewController.swift:
```diff
@@ -0,0 +1,141 @@
+                                                   categoryViewModel: categoryViewModel)
+                }
+                index += 1
+            }
```
이것도 하나의 유즈케이스면 별도로 분리해도 되는게 아닐까요.

ehgud0670 replied:
넵 맞습니다, 하지만 응답 데이터를 받기만 하고 다른 모델 객체 같은 것을 만들지 않아서 
좀 뭔가 UseCase 객체로 만들면 안될 것 같아서 만들지 않았습니다. ㅜㅜ
반영하였습니다!

godrm commented on iOS/Sidedish/Sidedish/ViewControllers/CategoryViewController.swift:
```diff
@@ -0,0 +1,141 @@
+    }
+    
+    override func viewDidAppear(_ animated: Bool) {
+        presentLoginViewController()
```
override 하는 메소드는 super.xxx()를 호출하는 습관을 가지세요. (물론 아주 드물게 호출하지 않아도 되는 예외가 있긴 합니다만)
그렇지 않으면 제대로 동작하지 않는 경우가 많습니다.

ehgud0670 replied:
넵, 반영하였습니다. override 하는 메소드는 super를 호출하도록 하겠습니다
감사합니다.

godrm commented on iOS/Sidedish/Sidedish/ViewControllers/CategoryViewController.swift:
```diff
@@ -0,0 +1,141 @@
+        categoryViewModel.performBind { header in
+            categoryHeaderView.configureCategory(text: header.name)
+            categoryHeaderView.configureTitle(text: header.description)
+        }
```
이런거는 bind라고 하기는 그렇고, 그냥 `categoryHeaderView.configure(with: categoryViewModel.header)` 정도가 더 간단하지 않나요?
꼭 클로저로 해야하는 건가요?

ehgud0670 replied:
오 넵넵, header를 넘기는 간단한 경우로 바꾸었습니다. 감사합니다!

ehgud0670 committed:
[iOS] Refactor: 매개변수로 옵셔널이 필요없어서 non-optional 타입으로 바꿈
associate with #7

ehgud0670 committed:
[iOS] Refactor: override 하는 메소드라 super 메소드를 호출하였다.

ehgud0670 committed:
[iOS] Refactor: UseCase 객체가 모델을 만들고 모델을 핸들러로 던지면
	받은 부분에서 뷰모델 객체를 만들도록 기능구현험

ehgud0670 committed:
[iOS] Refactor: 반찬 섹션들에 대한 url 배열을 가져오는 UseCase 만들어서 재적용함

ehgud0670 committed:
[iOS] Refactor: header를 view 에 넘겨주는 방식이 더 간단하여 리팩토링함
	따라서 ViewModelBinding의 performBind가 필요없어져서 삭제함

ehgud0670 committed:
[iOS] Style: 모크와 스텁을 혼용해서 쓰지 않기 위해 이름에서 스텁을 삭제함
associate with 7

ehgud0670 committed:
[iOS] Refactor: 해당 메소드들은 정적 메소드가 적절하기에 변경함

ehgud0670 committed:
[iOS] Refactor: 뷰모델 객체가 데이터소스를 채택하는 형식으로 구조를 바꿈
	이유: 뷰모델이란 뷰를 그리기 위한 데이터를 가지고 있고, 테이블 뷰 데이터소스는
	      테이블 뷰를 그리기 위한 데이터를 갖고 있어야 하므로 서로 똑같은 일을 하고 있는 것이다.
		따라서 뷰모델 객체가 테이블 뷰 데이터소스의 역할도 하는 것이 적절하기 때문에 채택하도록 하였다.
